### PR TITLE
Minor React styleguide updates

### PR DESCRIPTION
**Pull request by lemonmade**
This PR updates a few things on the React styleguide:

- Added back a warning for binding in `render`. Given all the performance talk lately, it's nice to encourage not binding so that we don't create extra rendering and we give components the ability to optimize using shallow comparisons of props.

- Recommended importing the things you need from React (`Component`, `PropTypes`, etc), instead of accessing them as properties on `React`.

- Updated recommendation for how to manage components in a folder with multiple subcomponents — do `Component/Component.js`, `Component/Subcomponent.js`, and export using `Component/index.js`, so that you get good file names everywhere and can still get easy imports elsewhere.

cc/ @Shopify/fed @bouk @harismahmood89

### CODE REVIEW

**Comment by bouk** [comment]
Can we recommend people to just drop prop types and use flow instead? It's so, so much nicer

File: react/README.md [hunk 1]
```diff
@@ -498,22 +450,76 @@ This guide provides a few guidelines on writing sensible React. Many of these ru
...
+  }
+
+  // also good
+  class GoodComponent extends Component {
+    handleClick = this.handleClick.bind(this);
+    handleMouseEnter = this.handleMouseEnter.bind(this);
+
+    render() {
+      return <button onClick={this.handleClick} onMouseEnter={this.handleMouseEnter} />;
+    }
```
Comment (perrupa) [inline react/README.md]: did you mean to leave the `this.` in this last example? or did you mean:
```js
return <button onClick={handleClick} onMouseEnter={handleMouseEnter} />;

```

Comment (lemonmade) [inline react/README.md]: @perrupa nope, the properties above this are class instance properties, so they are still accessed with `this` (they just give you a quicker way of setting up instance properties that are constant expressions than to need to do it in the constructor).

Comment (perrupa) [inline react/README.md]: I guess I was expecting colons for assignment... My bad, carry on 😄

File: react/README.md [hunk 2]
```diff
@@ -26,48 +27,28 @@ This guide provides a few guidelines on writing sensible React. Many of these ru
 
   ```js
   // bad
-  class financialReport extends React.Component {}
+  class financialReport extends Component {}
```
Comment (bouk) [inline react/README.md]: why this change? The react guide does `React.Component`, I think we should just follow that by default https://facebook.github.io/react/docs/reusable-components.html#es6-classes

File: react/README.md [hunk 3]
```diff
@@ -498,22 +450,76 @@ This guide provides a few guidelines on writing sensible React. Many of these ru
...
+
+    render() {
+      return <button onClick={this.handleClick} onMouseEnter={this.handleMouseEnter} />;
+    }
+  }
+
+  // also good
+  class GoodComponent extends Component {
+    handleClick = this.handleClick.bind(this);
+    handleMouseEnter = this.handleMouseEnter.bind(this);
```
Comment (bouk) [inline react/README.md]: how about the following alternative:


```
handleMouseEnter = () => {
  ...
}
```

I'm quite a fan of it

File: react/README.md [hunk 4]
```diff
@@ -528,6 +534,51 @@ This guide provides a few guidelines on writing sensible React. Many of these ru
...
 
+## Modules
+
+- [7.1](#7.1) <a name="7.1"></a> When importing from React, explicitly import the exact features you need (in addition to React, which will be a signal to the compiler to compile JSX).
+
+  > Why? Named imports explicitly indicate what you will be using at the top of the file, and reduce the duplication of accessing properties on `React`.
+
+  ```js
+  // bad
+  import React from 'react';
```
Comment (bouk) [inline react/README.md]: this is how every react guide does it, I think it's fine

File: react/README.md [hunk 5]
```diff
@@ -528,6 +534,51 @@ This guide provides a few guidelines on writing sensible React. Many of these ru
...
+
+  ```js
+  // in Card/Section.js
+  export default class Section extends Component {}
+
+  // in Card/Card.js
+  import Section from './Section';
+
+  export default class Card extends Component {
+    static Section = Section;
```
Comment (bouk) [inline react/README.md]: I don't like this very much. I'd prefer renaming `Section` to `CardSection` so it's clear that it's related

File: react/README.md [hunk 6]
```diff
@@ -528,6 +534,51 @@ This guide provides a few guidelines on writing sensible React. Many of these ru
...
+  }
+
+  // in Card/index.js
+  import Card from './Card';
```
Comment (bouk) [inline react/README.md]: why not just name it `Card/index.js`? Pretty clear what it means.

File: react/README.md (hunk 5 above)
Comment (lemonmade) [inline react/README.md]: I saw that in your branch and I don't like the pattern — you end up having tons of components that are actually only ever supposed to be used as a child of a particular component but that isn't really explicit from the naming. I think `<Card.Section />` more clearly shows intent than `<CardSection />`, and is cleaner when there are lots of children components you might need.

File: react/README.md (hunk 6 above)
Comment (lemonmade) [inline react/README.md]: Because then all of the files you are editing are named `index.js` and it's harder to work in the project, I've been using this setup and it makes it much easier to write things (also keeps it consistent with components that are just <Component>.js and don't need their own directory.

File: react/README.md (hunk 4 above)
Comment (lemonmade) [inline react/README.md]: :-1:, when React inevitably does an ES6 version we will be happy that we are only importing what we need so that we can tree shake.

File: react/README.md (hunk 3 above)
Comment (lemonmade) [inline react/README.md]: I'm kind of torn on this, I like the consistency of the methods being all declared in the same way but this pattern is kind of annoyingly boilerplate-ey. Only real thing I can see is that doing it in the way you suggest creates a (probably?) larger reference that needs to be stored per instance, probably not a big deal though.

File: react/README.md (hunk 4 above)
Comment (bouk) [inline react/README.md]: all right I reverse my decision actually, the proper ES way of importing under namespace would be `import * as React from 'react'`, so what they're using in the guide is kind of outdated. nvm 👍

File: react/README.md (hunk 5 above)
Comment (bouk) [inline react/README.md]: If we're going to bring up tree shaking, this also makes it impossible to tree shake out any components that are assigned to the static of another like this

File: react/README.md (hunk 6 above)
Comment (bouk) [inline react/README.md]: not sure how the files being named index.js makes it harder, it's clear from the directory name what the file contains. It just adds clutter with an extra file that doesn't do much

Comment (lemonmade) [inline react/README.md]: I don't know what everyone's setup looks like, but in Sublime and Atom, files with the same name end up getting expanded out to show the full root-relative directory, which means you can maybe get a couple of `index.js` files open at a time before it becomes unreadable.

File: react/README.md (hunk 5 above)
Comment (lemonmade) [inline react/README.md]: The reason I think this is different is because when components are connected (like Card and CardSection), they are effectively guaranteed to be used together, so you won't be able to eliminate much if anything anyways.

File: react/README.md (hunk 6 above)
Comment (bouk) [inline react/README.md]: dunno if I did something funky with my sublime set up, but it only show the parent directory name next to it for me

File: react/README.md (hunk 5 above)
Comment (bouk) [inline react/README.md]: I just don't really like it because it kind of goes around the ES6 module system and depends on adding properties to things that, although they're related, don't _need_ that property. It feels like kind of a step backwards

Comment (lemonmade) [inline react/README.md]: Would it make sense to do them as named exports, so you would have the default be the "root" component and then named for each subcomponent? Personally I still like the static properties because they make it super clear in JSX that it's a child, but maybe named exports is a reasonable compromise?

Comment (bouk) [inline react/README.md]: yeah, I think it makes more sense as named exports

File: react/README.md (hunk 3 above)
Comment (lemonmade) [inline react/README.md]: @bouk how about we leave either one open (binding existing methods, or writing the methods as instance properties)?

File: react/README.md (hunk 2 above)
Comment (lemonmade) [inline react/README.md]: (discussed below)

File: react/README.md (hunk 3 above)
Comment (bouk) [inline react/README.md]: yeah I'm fine with both

**Comment by lemonmade** [comment]
@bouk updated to recommend Flow (allowing proptypes since some projects use them, and you need to use them for setting context types IIRC), and to encourage components that expose multiple subcomponents to do them through named exports. Thoughts?

**Comment by bouk** [comment]
:shipit:
